nixpkgs/nixpkgs.bzl

```diff
@@ -59,19 +93,196 @@ nixpkgs_local_repository = repository_rule(
+    # Because of https://github.com/bazelbuild/bazel/issues/5356 we can't
+    # directly pass a dict from strings to labels to the rule (which we'd like
+    # for the `repositories` arguments), but we can pass a dict from labels to
+    # strings. So we swap the keys and the values (assuming they all are
+    # distinct).
+    hasRepository = "repository" in kwargs and kwargs["repository"] != None
+    hasRepositories = "repositories" in kwargs and kwargs["repositories"] != None
+    if hasRepository and hasRepositories:
+        fail("Specify one of 'repository' or 'repositories' (but not both).")
+    if hasRepositories:
+        inversed_repositories = invert_dict(kwargs["repositories"])
+        kwargs["repositories"] = inversed_repositories
+    if hasRepository:
+        repository = kwargs.pop("repository")
+        kwargs["repositories"] = { repository: "nixpkgs" }
+    f(*args, **kwargs)
+
+def _readlink(repository_ctx, path):
+    return repository_ctx.path(path).realpath
 
+def _nixpkgs_packages_instantiate_impl(repository_ctx):
     # Is nix supported on this platform?
     not_supported = not _is_supported_platform(repository_ctx)
     # Should we fail if Nix is not supported?
     fail_not_supported = repository_ctx.attr.fail_not_supported
 
-    if repository and repositories or not repository and not repositories:
-        fail("Specify one of 'repository' or 'repositories' (but not both).")
-    elif repository:
-        repositories = {repository_ctx.attr.repository: "nixpkgs"}
+    repository_ctx.file("BUILD", """exports_files(glob(["*"]))""")
+
+    if not_supported and fail_not_supported:
+        fail("Platform is not supported (see 'fail_not_supported')")
+    elif not_supported:
+        return
+
+    nix_instantiate_path = _executable_path(
+        repository_ctx,
+        "nix-instantiate",
+        extra_msg = "See: https://nixos.org/nix/",
+    )
+
+    # Assumes `{name}` does not contain any `"` and `{attrPath}` uses a valid
+    # attribute path syntax
+    packages_from_attr = \
+        [
+            "\"{name}\" = nixpkgs.{attrPath};".format(name = name, attrPath = attrPath)
+            for (name, attrPath) in repository_ctx.attr.packagesFromAttr.items()
+        ]
+
+    generatedPackageFiles = {}
+    for (name, expr) in repository_ctx.attr.packagesFromExpr.items():
+        nix_definition_file = "__internal_" + name + "_definition.nix"
+        repository_ctx.file(
+            nix_definition_file,
+            content = expr
+            )
+        generatedPackageFiles[name] = "./" + nix_definition_file
+    packagesFromFileLabels = \
+      { name: repository_ctx.path(Label(filePath))
+        for (name, filePath) in repository_ctx.attr.packagesFromFile.items()
+      }
+    packagesFromFile = packagesFromFileLabels + generatedPackageFiles
+
+    all_packages = _implode_packages_defs(
+        packagesFromFile,
+        repository_ctx.attr.packagesFromAttr,
+    )
```

Comment:
The logic here from line 170 is a bit non-self-explanatory, I feel like it needs some comments.

Might be a good thing to have for the whole subroutine.